File: frontend/src/lib/config-parse.ts
/** Parse / format dashboard pseudo-Python config (mirrors backend config_parse.py). */

export const SEARCH_LIST_KEYS = new Set([
  "search_terms",
  "experience_level",
  "job_type",
  "on_site",
  "companies",
  "location",
  "industry",
  "job_function",
  "job_titles",
  "benefits",
  "commitments",
  "about_company_bad_words",
  "about_company_good_words",
  "bad_words",
]);

function tryParseList(text: string): string[] | null {
  const trimmed = text.trim();
  if (!trimmed.startsWith("[")) return null;
  try {
    const normalized = trimmed.replace(/'/g, '"');
    const parsed = JSON.parse(normalized);
    return Array.isArray(parsed) ? parsed.map(String) : null;
  } catch {
    return null;
  }
}

export function normalizeConfigValue(key: string, value: unknown): unknown {
  if (value === null || value === undefined) {
    return SEARCH_LIST_KEYS.has(key) ? [] : "";
  }
  if (Array.isArray(value)) {
    return value;
  }
  if (typeof value === "string") {
    const text = value.trim();
    if (SEARCH_LIST_KEYS.has(key)) {
      if (text === "[" || text === '["' || text === "']" || text === "") {
        return [];
      }
      const asList = tryParseList(text);
      if (asList) return asList;
      return text ? [text] : [];
    }
  }
  return value;
}

export function parseConfigValue(raw: string, key = ""): unknown {
  const valueStr = raw.trim();
  if (!valueStr) return "";

  if (
    (valueStr.startsWith('"') && valueStr.endsWith('"')) ||
    (valueStr.startsWith("'") && valueStr.endsWith("'"))
  ) {
    const inner = valueStr
      .slice(1, -1)
      .replace(/\\n/g, "\n")
      .replace(/\\"/g, '"');
    if (SEARCH_LIST_KEYS.has(key) || inner.trim().startsWith("[")) {
      const asList = tryParseList(inner);
      if (asList) return asList;
    }
    if (SEARCH_LIST_KEYS.has(key) && (inner.trim() === "[" || inner.trim() === "")) {
      return [];
    }
    return inner;
  }

  const lower = valueStr.toLowerCase();
  if (lower === "true") return true;
  if (lower === "false") return false;

  const asList = tryParseList(valueStr);
  if (asList) return asList;

  if (valueStr === "[" || valueStr === '["') return [];

  if (/^-?\d+$/.test(valueStr)) return Number(valueStr);
  if (/^-?\d+\.\d+$/.test(valueStr)) return Number(valueStr);

  if (SEARCH_LIST_KEYS.has(key)) {
    return valueStr ? [valueStr] : [];
  }
  return valueStr;
}

export function parseConfigContent(content: string): Record<string, unknown> {
  const parsed: Record<string, unknown> = {};
  const lines = content.split("\n");
  let currentKey = "";
  let currentValue = "";
  let inQuotedString = false;
  let quoteChar = '"';

  for (const line of lines) {
    if (!inQuotedString) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith("#")) continue;
      if (!line.includes("=")) continue;

      const eq = line.indexOf("=");
      currentKey = line.slice(0, eq).trim();
      const val = line.slice(eq + 1).trim();

      if (val.startsWith('"') || val.startsWith("'")) {
        quoteChar = val[0];
        inQuotedString = true;
        currentValue = val.slice(1);
        if (currentValue.endsWith(quoteChar) && currentValue.length > 0) {
          inQuotedString = false;
          parsed[currentKey] = parseConfigValue(
            quoteChar + currentValue.slice(0, -1) + quoteChar,
            currentKey
          );
        }
      } else {
        parsed[currentKey] = parseConfigValue(val, currentKey);
      }
    } else if (line.endsWith(quoteChar)) {
      inQuotedString = false;
      currentValue += "\n" + line.slice(0, -1);
      parsed[currentKey] = parseConfigValue(
        quoteChar + currentValue + quoteChar,
        currentKey
      );
    } else {
      currentValue += "\n" + line;
    }
  }

  if (inQuotedString && currentKey) {
    parsed[currentKey] = parseConfigValue(
      quoteChar + currentValue + quoteChar,
      currentKey
    );
  }

  for (const [key, value] of Object.entries(parsed)) {
    parsed[key] = normalizeConfigValue(key, value);
  }

  return parsed;
}

export function formatConfigValue(value: unknown): string {
  if (typeof value === "boolean") {
    return value ? "True" : "False";
  }
  if (Array.isArray(value)) {
    return JSON.stringify(value).replace(/"/g, "'");
  }
  if (typeof value === "string") {
    const escaped = value
      .replace(/\\/g, "\\\\")
      .replace(/"/g, '\\"')
      .replace(/\n/g, "\\n");
    return `"${escaped}"`;
  }
  if (value === null || value === undefined) {
    return '""';
  }
  return String(value);
}

export function formatConfigContent(
  category: string,
  data: Record<string, unknown>
): string {
  let out = `################ ${category.toUpperCase()} CONFIGURATION ################\n\n`;
  for (const [key, value] of Object.entries(data)) {
    const normalized = normalizeConfigValue(key, value);
    out += `${key} = ${formatConfigValue(normalized)}\n`;
  }
  return out;
}

File: backend/routes/config_parse.py
"""Parse and format dashboard pseudo-Python config files."""

from __future__ import annotations

import ast
import json
import re

# Keys that must be lists for the job-applier validator (search category).
SEARCH_LIST_KEYS = frozenset(
    {
        "search_terms",
        "experience_level",
        "job_type",
        "on_site",
        "companies",
        "location",
        "industry",
        "job_function",
        "job_titles",
        "benefits",
        "commitments",
        "about_company_bad_words",
        "about_company_good_words",
        "bad_words",
    }
)


def _try_literal_list(value: str):
    text = value.strip()
    if not text.startswith("["):
        return None
    try:
        parsed = ast.literal_eval(text)
    except (SyntaxError, ValueError):
        return None
    return parsed if isinstance(parsed, list) else None


def normalize_stored_value(key: str, value):
    """Repair corrupted DB values (e.g. literal ``'['``) before export to the UI."""
    if value is None:
        return [] if key in SEARCH_LIST_KEYS else ""
    if isinstance(value, list):
        return value
    if not isinstance(value, str):
        return value

    text = value.strip()
    if key in SEARCH_LIST_KEYS:
        if text in ("[", '["', "']", '"[]"'):
            return []
        as_list = _try_literal_list(text)
        if as_list is not None:
            return as_list
        if text:
            return [text]
        return []
    return value


def parse_config_value(value_str: str, key: str = ""):
    """Parse one RHS from ``key = value`` in a config file."""
    value_str = value_str.strip()
    if not value_str:
        return ""

    if (value_str.startswith('"') and value_str.endswith('"')) or (
        value_str.startswith("'") and value_str.endswith("'")
    ):
        inner = value_str[1:-1]
        inner = inner.replace("\\n", "\n").replace('\\"', '"')
        if key in SEARCH_LIST_KEYS or inner.strip().startswith("["):
            as_list = _try_literal_list(inner)
            if as_list is not None:
                return as_list
        if key in SEARCH_LIST_KEYS and inner.strip() in ("[", ""):
            return []
        return inner

    lowered = value_str.lower()
    if lowered == "true":
        return True
    if lowered == "false":
        return False

    as_list = _try_literal_list(value_str)
    if as_list is not None:
        return as_list

    if value_str in ("[", '["'):
        return []

    if re.fullmatch(r"-?\d+", value_str):
        return int(value_str)
    try:
        return float(value_str)
    except ValueError:
        pass

    if key in SEARCH_LIST_KEYS:
        return [] if not value_str else [value_str]
    return value_str


def format_config_value(value) -> str:
    """Serialize a Python value for pseudo-config file content."""
    if isinstance(value, bool):
        return "True" if value else "False"
    if isinstance(value, list):
        return json.dumps(value).replace('"', "'")
    if isinstance(value, str):
        escaped = value.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n")
        return f'"{escaped}"'
    if value is None:
        return '""'
    return str(value)


def parse_config_content(content: str) -> dict:
    """Parse a full config file into a dict (supports multi-line quoted strings)."""
    parsed: dict = {}
    lines = content.split("\n")
    current_key = ""
    current_value = ""
    in_quoted_string = False
    quote_char = '"'

    for line in lines:
        if not in_quoted_string:
            trimmed = line.strip()
            if not trimmed or trimmed.startswith("#"):
                continue
            if "=" not in line:
                continue
            key, val = line.split("=", 1)
            current_key = key.strip()
            val = val.strip()

            if val.startswith('"') or val.startswith("'"):
                quote_char = val[0]
                in_quoted_string = True
                current_value = val[1:]
                if current_value.endswith(quote_char) and len(current_value) > 0:
                    in_quoted_string = False
                    parsed[current_key] = parse_config_value(
                        quote_char + current_value[:-1] + quote_char, current_key
                    )
            else:
                parsed[current_key] = parse_config_value(val, current_key)
        else:
            if line.endswith(quote_char):
                in_quoted_string = False
                current_value += "\n" + line[: -len(quote_char)]
                parsed[current_key] = parse_config_value(
                    quote_char + current_value + quote_char, current_key
                )
            else:
                current_value += "\n" + line

    if in_quoted_string and current_key:
        parsed[current_key] = parse_config_value(
            quote_char + current_value + quote_char, current_key
        )

    return parsed


def format_config_content(category: str, config_data: dict) -> str:
    header = f"################ {category.upper()} CONFIGURATION ################\n\n"
    lines = [header]
    for key, value in config_data.items():
        if category == "search":
            value = normalize_stored_value(key, value)
        lines.append(f"{key} = {format_config_value(value)}\n")
    return "".join(lines)
